### PR TITLE
feat: Azure DataLake compatibility + metadata addition

### PR DESCRIPTION
**Pull request by ank19**
## What this PR changes/adds

This PR adds compatibility for the Azure DataLake including federated identity authentication. For convenience, it also adds metadata to the uploaded blob objects.

## Why it does that

1. Currently, the extension only supports authentication by providing a shared key. Explicit credential management for companies implies efforts and costs, adhering to security / compliance rules like key rotation. With a few lines of code, a huge security and operational efficiency can be achieved by supporting a standard Microsoft feature named "Federated Identity" for newer services like Azure Data Lake Gen2.
2. Currently, data received by the extension sort-of "just appears" in the storage account and has to be somehow correlated to what it belongs to. Adding metadata would solve this in context of this existing extension, especially adding e. g. adding the BPN of the sending business partner and adding a correlation id, which can be supplied in the data destination of the transfer request (to be discussed).
3. Currently, the implementation is focused on the old Azure Storage account, specifically it does not support directories of the successor service, Azure Data Lake Gen2, and it would be most likely of interest to all parties integrating the EDC on Azure to use this feature, as it has especially a significant cost impact, due to the possibility to selectively scan the Azure Data Lake based on directories.

## Further notes

Concerning 1) basically the change is just about attempting to instantiate the DefaultAzureCredentialBuilder instead of using explicit keys if the no credentials are provided (see createSource in AzureStorageDataSourceFactory); implying that these are not mandatory anymore
Concerning 2) I tried stick to the existing code style (decorator, see org.eclipse.edc.connector.dataplane.azure.storage.metadata)
Concerning 3) That just about adding some logic to determine the destination name (AzureStorageDataSink)

## Linked Issue(s)

- https://github.com/eclipse-edc/Technology-Azure/issues/75
- https://github.com/eclipse-edc/Technology-Azure/issues/74
- https://github.com/eclipse-edc/Technology-Azure/issues/73

### CODE REVIEW

**Comment by ank19** [comment]
@paullatzelsperger - This PR replaces the [old one](https://github.com/eclipse-edc/Technology-Azure/pull/71) - unfortunately my git config got lost, so I had to recreate it. Sorry for the circumstances.

**Comment by paullatzelsperger** [comment]
> @paullatzelsperger - This PR replaces the [old one](https://github.com/eclipse-edc/Technology-Azure/pull/71) - unfortunately my git config got lost, so I had to recreate it. Sorry for the circumstances.

no worries. I requested reviews of the technical committers.

File: extensions/common/azure/azure-blob-core/src/main/java/org/eclipse/edc/azure/blob/api/BlobStoreApiImpl.java
Comment (bscholtes1A) [inline extensions/common/azure/azure-blob-core/src/main/java/org/eclipse/edc/azure/blob/api/BlobStoreApiImpl.java:115]: final to be removed (see guidelines)

File: extensions/common/azure/azure-blob-core/src/main/java/org/eclipse/edc/azure/blob/api/BlobStoreApiImpl.java
Comment (bscholtes1A) [inline extensions/common/azure/azure-blob-core/src/main/java/org/eclipse/edc/azure/blob/api/BlobStoreApiImpl.java:141]: final

File: extensions/common/azure/azure-blob-core/src/main/java/org/eclipse/edc/azure/blob/api/BlobStoreApiImpl.java
Comment (bscholtes1A) [inline extensions/common/azure/azure-blob-core/src/main/java/org/eclipse/edc/azure/blob/api/BlobStoreApiImpl.java:142]: var

File: extensions/data-plane/data-plane-azure-storage/src/main/java/org/eclipse/edc/connector/dataplane/azure/storage/metadata/BlobMetadataProviderImpl.java
Comment (bscholtes1A) [inline extensions/data-plane/data-plane-azure-storage/src/main/java/org/eclipse/edc/connector/dataplane/azure/storage/metadata/BlobMetadataProviderImpl.java:26]: `blobMetadataDecorators` instead of `sinkDecorators` ?

File: extensions/data-plane/data-plane-azure-storage/src/main/java/org/eclipse/edc/connector/dataplane/azure/storage/metadata/BlobMetadataProvider.java
Comment (bscholtes1A) [inline extensions/data-plane/data-plane-azure-storage/src/main/java/org/eclipse/edc/connector/dataplane/azure/storage/metadata/BlobMetadataProvider.java:22]: registerDecorator

File: extensions/data-plane/data-plane-azure-storage/src/main/java/org/eclipse/edc/connector/dataplane/azure/storage/metadata/CommonBlobMetadataDecorator.java
Comment (bscholtes1A) [inline extensions/data-plane/data-plane-azure-storage/src/main/java/org/eclipse/edc/connector/dataplane/azure/storage/metadata/CommonBlobMetadataDecorator.java:49]: See `isNullOrEmpty` of `StringUtils`

File: extensions/data-plane/data-plane-azure-storage/src/main/java/org/eclipse/edc/connector/dataplane/azure/storage/metadata/CommonBlobMetadataDecorator.java
Comment (bscholtes1A) [inline extensions/data-plane/data-plane-azure-storage/src/main/java/org/eclipse/edc/connector/dataplane/azure/storage/metadata/CommonBlobMetadataDecorator.java:47]: Remove these final

File: extensions/data-plane/data-plane-azure-storage/src/main/java/org/eclipse/edc/connector/dataplane/azure/storage/pipeline/AzureStorageDataSink.java
Comment (bscholtes1A) [inline extensions/data-plane/data-plane-azure-storage/src/main/java/org/eclipse/edc/connector/dataplane/azure/storage/pipeline/AzureStorageDataSink.java:59]: final

File: extensions/data-plane/data-plane-azure-storage/src/main/java/org/eclipse/edc/connector/dataplane/azure/storage/pipeline/AzureStorageDataSink.java
Comment (bscholtes1A) [inline extensions/data-plane/data-plane-azure-storage/src/main/java/org/eclipse/edc/connector/dataplane/azure/storage/pipeline/AzureStorageDataSink.java:89]: final

File: extensions/data-plane/data-plane-azure-storage/src/main/java/org/eclipse/edc/connector/dataplane/azure/storage/pipeline/AzureStorageDataSink.java
Comment (bscholtes1A) [inline extensions/data-plane/data-plane-azure-storage/src/main/java/org/eclipse/edc/connector/dataplane/azure/storage/pipeline/AzureStorageDataSink.java:113]: final

Comment (bscholtes1A) [inline extensions/data-plane/data-plane-azure-storage/src/main/java/org/eclipse/edc/connector/dataplane/azure/storage/pipeline/AzureStorageDataSink.java:113]: Use isNullOrEmpty from StringUtils

File: extensions/data-plane/data-plane-azure-storage/src/main/java/org/eclipse/edc/connector/dataplane/azure/storage/pipeline/AzureStorageDataSource.java
Comment (bscholtes1A) [inline extensions/data-plane/data-plane-azure-storage/src/main/java/org/eclipse/edc/connector/dataplane/azure/storage/pipeline/AzureStorageDataSource.java:61]: Cant we use a Builder pattern for AzureStoragePart? To avoid this if condition

File: extensions/data-plane/data-plane-azure-storage/src/test/java/org/eclipse/edc/connector/dataplane/azure/storage/AzureDataPlaneCopyIntegrationTest.java
Comment (bscholtes1A) [inline extensions/data-plane/data-plane-azure-storage/src/test/java/org/eclipse/edc/connector/dataplane/azure/storage/AzureDataPlaneCopyIntegrationTest.java:138]: final

File: extensions/data-plane/data-plane-azure-storage/src/test/java/org/eclipse/edc/connector/dataplane/azure/storage/metadata/BlobMetadataProviderImplTest.java
Comment (bscholtes1A) [inline extensions/data-plane/data-plane-azure-storage/src/test/java/org/eclipse/edc/connector/dataplane/azure/storage/metadata/BlobMetadataProviderImplTest.java:34]: final again

File: extensions/data-plane/data-plane-azure-storage/src/main/java/org/eclipse/edc/connector/dataplane/azure/storage/metadata/CommonBlobMetadataDecorator.java
Comment (wolf4ood) [inline extensions/data-plane/data-plane-azure-storage/src/main/java/org/eclipse/edc/connector/dataplane/azure/storage/metadata/CommonBlobMetadataDecorator.java:44]:
```suggestion
        builder.put("originalName", part.name())
                .put("requestId", request.getId())
                .put("processId", request.getProcessId())
                .put("connectorId", context.getConnectorId())
                .put("participantId", context.getParticipantId());
```

File: extensions/data-plane/data-plane-azure-storage/src/test/java/org/eclipse/edc/connector/dataplane/azure/storage/metadata/CommonBlobMetadataDecoratorTest.java
Comment (wolf4ood) [inline extensions/data-plane/data-plane-azure-storage/src/test/java/org/eclipse/edc/connector/dataplane/azure/storage/metadata/CommonBlobMetadataDecoratorTest.java:37]:
```suggestion
    private final TypeManager typeManager = mock();
```

File: extensions/data-plane/data-plane-azure-storage/src/test/java/org/eclipse/edc/connector/dataplane/azure/storage/pipeline/AzureDataSourceToDataSinkTest.java
Comment (wolf4ood) [inline extensions/data-plane/data-plane-azure-storage/src/test/java/org/eclipse/edc/connector/dataplane/azure/storage/pipeline/AzureDataSourceToDataSinkTest.java:222]: remove final please

File: extensions/data-plane/data-plane-azure-storage/src/test/java/org/eclipse/edc/connector/dataplane/azure/storage/pipeline/AzureStorageDataSinkFactoryTest.java
Comment (wolf4ood) [inline extensions/data-plane/data-plane-azure-storage/src/test/java/org/eclipse/edc/connector/dataplane/azure/storage/pipeline/AzureStorageDataSinkFactoryTest.java:45]: You could use type inference here
```suggestion
    private final ServiceExtensionContext context = mock();
```

File: extensions/data-plane/data-plane-azure-storage/src/test/java/org/eclipse/edc/connector/dataplane/azure/storage/pipeline/AzureStorageDataSinkTest.java
Comment (wolf4ood) [inline extensions/data-plane/data-plane-azure-storage/src/test/java/org/eclipse/edc/connector/dataplane/azure/storage/pipeline/AzureStorageDataSinkTest.java:63]: also here a nit of type inference

**Review by wolf4ood — COMMENTED** [review]
Some minor improvement suggestion in addition to @bscholtes1A review

**Comment by paullatzelsperger** [comment]
@ank19 please find [here](https://github.com/eclipse-edc/docs/tree/main/developer/contributing) the coding style guide, which must be adhered to. They are referenced in our [PR etiquette](https://github.com/eclipse-edc/docs/blob/main/contributing/pr_etiquette.md), which in turn is mentioned in the contributing guidelines. Those are accessible from the [project landing page](https://github.com/eclipse-edc). 

Once you've applied them all, please re-request a review from committers, who already reviewed the PR.

**Comment by ank19** [comment]
@paullatzelsperger - Thanks for the references.

I committed the changes according to the review comments, with the exception of one open query/counter-suggestion regarding the creation of a builder from @bscholtes1A ; I will care about it after getting feedback regarding this.

Essentially
- Removal of "final" for local variables
- Removal of explicit paramaters for mock() where possible
- "var" where possible
The existing code, insofar as it is part of the Azure Blob extension or it's dependencies, I have also adjusted concerning these aspects.

**Review by bscholtes1A — APPROVED** [review]

File: extensions/data-plane/data-plane-azure-storage/src/main/java/org/eclipse/edc/connector/dataplane/azure/storage/pipeline/AzureStorageDataSource.java
Comment (ank19) [reply extensions/data-plane/data-plane-azure-storage/src/main/java/org/eclipse/edc/connector/dataplane/azure/storage/pipeline/AzureStorageDataSource.java:61]: @bscholtes1A - Yes, I totally agree with you about that BlobStoreApi, I struggled with that one, too. I just got a message that the code coverage is down by 1% (new code). I guess one reason might be that I changed everything to "var", even if it does not touch the actual functionality of this PR, and the BlobStoreApi is covered only 1/3. I can offer to write a unit test for that part, but if you want to refactor it anyhow, that would not make sense. Just let me know if you want me to do so anyhow. Regarding your proposal: I applied your suggestions but did not commit it yet, just waiting for other reviewers.

**Comment by bscholtes1A** [comment]
@ank19 did you processed all comments from @wolf4ood  and myself? If yes I think we are good to go once you fix the dependencies. 
We can tackle the refactoring of the BlobStoreApi in a second step.

**Comment by ank19** [comment]
@bscholtes1A - I processed all the comments, yes. I just added changes according to your latest comments. As I'm unfortunately not familiar with the environment, can you give me a hint what to do regarding the dependencies? It there a script to execute?

**Comment by paullatzelsperger** [comment]
> @bscholtes1A - I processed all the comments, yes. I just added changes according to your latest comments. As I'm unfortunately not familiar with the environment, can you give me a hint what to do regarding the dependencies? It there a script to execute?

it's a bit of a chore, but unfortunately its a necessary one. if you check the failing [job](https://github.com/eclipse-edc/Technology-Azure/actions/runs/6571332321/job/17863355596), it should basically tell you exactly what needs to be in the `DEPENDENCIES` file. Don't forget to add a newline at the end of the file.

File: extensions/common/azure/azure-blob-core/src/main/java/org/eclipse/edc/azure/blob/api/BlobStoreApi.java
Comment (paullatzelsperger) [inline extensions/common/azure/azure-blob-core/src/main/java/org/eclipse/edc/azure/blob/api/BlobStoreApi.java:48]: some documentation would be nice. I know, the rest of the interface isn't documented either, but maybe you could put a comment on this added method (e.g. why there is no auth parameter)

File: extensions/data-plane/data-plane-azure-storage/build.gradle.kts
Comment (paullatzelsperger) [inline extensions/data-plane/data-plane-azure-storage/build.gradle.kts:29]: couldn't this be just `implementation`? (same for the next)

File: extensions/data-plane/data-plane-azure-storage/src/main/java/org/eclipse/edc/connector/dataplane/azure/storage/metadata/BlobMetadataDecorator.java
Comment (paullatzelsperger) [inline extensions/data-plane/data-plane-azure-storage/src/main/java/org/eclipse/edc/connector/dataplane/azure/storage/metadata/BlobMetadataDecorator.java:20]: doc pls

File: extensions/data-plane/data-plane-azure-storage/src/main/java/org/eclipse/edc/connector/dataplane/azure/storage/metadata/BlobMetadataProvider.java
Comment (paullatzelsperger) [inline extensions/data-plane/data-plane-azure-storage/src/main/java/org/eclipse/edc/connector/dataplane/azure/storage/metadata/BlobMetadataProvider.java:20]: doc pls

File: extensions/data-plane/data-plane-azure-storage/src/main/java/org/eclipse/edc/connector/dataplane/azure/storage/metadata/CommonBlobMetadataDecorator.java
Comment (paullatzelsperger) [inline extensions/data-plane/data-plane-azure-storage/src/main/java/org/eclipse/edc/connector/dataplane/azure/storage/metadata/CommonBlobMetadataDecorator.java:38]: no magic strings pls. use constants.

File: extensions/data-plane/data-plane-azure-storage/src/main/java/org/eclipse/edc/connector/dataplane/azure/storage/metadata/CommonBlobMetadataDecorator.java
Comment (paullatzelsperger) [inline extensions/data-plane/data-plane-azure-storage/src/main/java/org/eclipse/edc/connector/dataplane/azure/storage/metadata/CommonBlobMetadataDecorator.java:48]: this should probably be taken from the `AzureBlobSchema`

File: extensions/data-plane/data-plane-azure-storage/src/test/java/org/eclipse/edc/connector/dataplane/azure/storage/metadata/BlobMetadataProviderImplTest.java
Comment (paullatzelsperger) [inline extensions/data-plane/data-plane-azure-storage/src/test/java/org/eclipse/edc/connector/dataplane/azure/storage/metadata/BlobMetadataProviderImplTest.java:32]:
```suggestion
        var provider = new BlobMetadataProviderImpl(mock());
```

File: extensions/data-plane/data-plane-azure-storage/src/test/java/org/eclipse/edc/connector/dataplane/azure/storage/metadata/BlobMetadataProviderImplTest.java
Comment (paullatzelsperger) [inline extensions/data-plane/data-plane-azure-storage/src/test/java/org/eclipse/edc/connector/dataplane/azure/storage/metadata/BlobMetadataProviderImplTest.java:33]: pls do not mock value objects. either use a dummy (private nested class), or an actual implementation, e.g. `AzureStoragePart`.

File: extensions/data-plane/data-plane-azure-storage/src/test/java/org/eclipse/edc/connector/dataplane/azure/storage/metadata/BlobMetadataTest.java
Comment (paullatzelsperger) [inline extensions/data-plane/data-plane-azure-storage/src/test/java/org/eclipse/edc/connector/dataplane/azure/storage/metadata/BlobMetadataTest.java:46]: weird formatting

Comment (paullatzelsperger) [inline extensions/data-plane/data-plane-azure-storage/src/test/java/org/eclipse/edc/connector/dataplane/azure/storage/metadata/BlobMetadataTest.java:46]: pls use AssertJ assertions everywhere. E.g. `assertThatThrownBy` instead of `assertThrows`

File: extensions/data-plane/data-plane-azure-storage/src/test/java/org/eclipse/edc/connector/dataplane/azure/storage/metadata/BlobMetadataTest.java
Comment (paullatzelsperger) [inline extensions/data-plane/data-plane-azure-storage/src/test/java/org/eclipse/edc/connector/dataplane/azure/storage/metadata/BlobMetadataTest.java:57]: could be fluent

**Review by paullatzelsperger — COMMENTED** [review]
A few minor things. Please be sure to:
- **not** mock value objects
- properly name test methods, cf. [coding principles](https://github.com/eclipse-edc/docs/blob/main/developer/contributing/coding-principles.md#v-testing)
- document interfaces and methods

File: extensions/common/azure/azure-blob-core/src/main/java/org/eclipse/edc/azure/blob/api/BlobStoreApi.java
Comment (ank19) [reply extensions/common/azure/azure-blob-core/src/main/java/org/eclipse/edc/azure/blob/api/BlobStoreApi.java:81]: Sure, no problem. I documented the three getBlobAdapter methods to illustrate the differences. I would've documented the others as well, but as discussed with @bscholtes1A, I think a refactoring would make sense here and I did not want to spent unnecessary efforts. Hope that's fine.

**Comment by ank19** [comment]
@paullatzelsperger - Thanks for the review and for the hint with the dependencies. I saw the log entry in the middle of the workflow log for the dependencies to copy the output, so I will to follow that after the next run of the workflow. Concerning the changes: I applied them and committed them just now.

**Comment by ank19** [comment]
@bscholtes1A - Just done that for Paul's review; I was used to reviewers closing it in case they are not content with the changes... But either way is fine for me.

**Review by paullatzelsperger — APPROVED** [review]
LGTM!

**Comment by ank19** [comment]
@paullatzelsperger - Thanks. I updated the DEPENDENCIES accordingly and checked by manually executing that dash.jar that there are no differences. Can you pls. approve another workflow run? Thanks!

**Pull request merged**
